Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+namespace Neos\Fusion\FusionObjects\Helpers;
+
+use Neos\Flow\Annotations as Flow;
+use Neos\Fusion\Core\Runtime;
+
+/**
+ * @Flow\Proxy(false)
+ */
+final class LazyProps implements \ArrayAccess, \Iterator
+{
+
+    /**
+     * @var array
+     */
+    private $valueCache = [];
+
+    /**
+     * @var string
+     */
+    private $parentPath;
+
+    /**
+     * @var Runtime
+     */
+    private $runtime;
+
+    /**
+     * Index of keys
+     *
+     * @var array
+     */
+    private $keys;
+
+    /**
+     * @var object
+     */
+    private $fusionObject;
+
+    /**
+     * @var array
+     */
+    private $effectiveContext;
+
+    public function __construct(
+        object $fusionObject,
+        string $parentPath,
+        Runtime $runtime,
+        array $keys,
+        array $effectiveContext
+    ) {
+        $this->fusionObject = $fusionObject;
+        $this->parentPath = $parentPath;
+        $this->runtime = $runtime;
+        $this->keys = array_flip($keys);
+        $this->effectiveContext = $effectiveContext;
+    }
+
+    public function offsetExists($path)
+    {
+        return array_key_exists($path, $this->keys);
+    }
+
+    public function offsetGet($path)
+    {
+        if (!array_key_exists($path, $this->valueCache)) {
+            $this->runtime->pushContextArray($this->effectiveContext);
+            try {
+                $this->valueCache[$path] = $this->runtime->evaluate($this->parentPath . '/' . $path, $this->fusionObject);
+            } finally {
+                $this->runtime->popContext();
+            }
+        }
+        return $this->valueCache[$path];
+    }
+
+    public function offsetSet($path, $value)
+    {
+        // NOOP
```

Comment:
Shouldn't that throw?

Reply:
You mean the "set call" should throw an exception? Yes, we could do that.

Reply:
I _still_ think trying to set a lazy prop should throw :)

Reply:
@albe It's a good idea to resolve all conversations before hitting the merge button. Otherwise open discussions might get lost – like this one.. :)